devito/at_controller.py

```diff
@@ -0,0 +1,344 @@
+            if "time" in line or "space" in line:  # ignores the first line of the report file
+                continue
+
+            split = line.split(' ')
+            if int(split[0]) == time_order and int(split[1]) == space_order:  # finds the one we are looking for
+
+                # Splits, converts all block sizes into int and returns
+                return [int(element) for element in split[2].split(',')]
+
+    return None  # returns none if no matching time/space order was found
+
+
+def get_optimal_block_size(shape, time_order, space_order):
+    """Gets optimal block size. Based on architecture Currently works only on linux
+
+    Args:
+        shape (tuple|list): shape of kernel
+        time_order (int): time order of kernel
+        space_order (int): space order of kernel
+
+    Returns:
+        int: most optimal size  for the block
+    """
+
+    # list [time_order][space_order] hardcoded number of loads from the roof line at: (change if necessary)
+    # https://docs.google.com/spreadsheets/d/1OmvsTftH3uCfYZj-1Lb-5Ji7edrURf0UzzywYaw0-FY/edit?ts=5745964f#gid=0
+    number_of_loads = [[11, 17, 23, 29, 35, 41, 47, 53], [19, 25, 31, 37, 43, 49, 55]]  # works only till space order 16
+
+    cache_s = int(cpuinfo.get_cpu_info()['l2_cache_size'].split(' ')[0])  # cache size
+    core_c = cpuinfo.get_cpu_info()['count']  # number of cores
+    loads_c = number_of_loads[time_order / 2][space_order / 2]  # number of loads
+
+    # ((C size / cores) / (4 * length inner most * kernel loads)
+    optimal_b_size = math.sqrt(((1000 * cache_s) / core_c) / (4 * shape[len(shape) - 1] * loads_c))
+    return int(round(optimal_b_size))  # rounds to the nearest integer
+
+
+class AtController(object):
+    """Class responsible for controlling auto tuning process, copying relevant files and collecting data"""
+
+    def __init__(self, isat_dir_path):
+        """Initialises at controller. Creates at working dir. Creates all isat-files
+
+        Args:
+            isat_dir_path (str): full path to isat installation directory
+
+        Raises:
+            ValueError: if isat installation directory does not exist
+        """
+        if not os.path.exists(isat_dir_path):
+            raise ValueError("isat installation directory does not exits")
+
+        self.time_order = None
+        self.space_order = None
+
+        self._isat_py_path = os.path.join(isat_dir_path, "Source/Python/isat.py")  # path to isat script that does at
+        self._at_work_dir = os.path.join(isat_dir_path, "Auto-tuning")  # working dir for at
+
+        at_file_dir = os.path.join(self._at_work_dir, "Isat-files")  # dir for keeping files necessary for at
+
+        # all these files are necessary for auto tuning process and they have to be in same dir as tuned file
+        # do not change these names
+        self._isat_build_f = "isat-build"
+        self._isat_clean_f = "isat-clean"
+        self._isat_test_f = "isat-test"
+        self._make_f = "makefile"
```

Comment:
From [here](https://www.gnu.org/software/make/manual/html_node/Makefile-Names.html):
``We recommend Makefile because it appears prominently near the beginning of a directory listing, right near other important files such as README.``